### PR TITLE
#2481 Report ignored source properties which are missing

### PR DESCRIPTION
**Pull request by incaseoftrouble**
PR for #2481 - WIP

Optionally get warned / error if ignored source properties are missing (this way you can ensure an exact 1:1 correspondence between the declarations and generated code)

### CODE REVIEW

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @incaseoftrouble. 

It is good. However, I think that we do not need to have a new option in the `@Mapper` or `@MapperConfig` for this. I think that we can immediately have a compile error if the user used an unknown source property in `BeanMapping#ignoreUnmappedSourceProperties`. The same error happens if you use `@Mapping(target = "unknown", ignore = true)`.

I know that this might lead to compile errors in existing code. However, in my opinion that is OK.

What do you think @sjaakd?

**Comment by incaseoftrouble** [comment]
> I know that this might lead to compile errors in existing code. However, in my opinion that is OK.

Ok, that's for you guys to decide 👍 I'll wait for your comments and adapt accordingly.

**Comment by filiphr** [comment]
Thanks for your understanding @incaseoftrouble.

An additional comment for why the tests are failing. We have integration tests that are taking all tests from our test suite and compile the (the so called fullFeatureTest). If a certain test has a compile error then the fullFeatureTest will fail.

We do have mappers that lead to compilation errors, as we want to test this. In order to satisfy the integration test such mappers have to contain `Erroneous` in their name. Classes with that in their name will be ignored.

Long story short, once we agree whether we need the new option or not (I am for not) we would need to change the mappers that test the compile errors to have `Erroneous` in their name

**Comment by sjaakd** [comment]
> It is good. However, I think that we do not need to have a new option in the `@Mapper` or `@MapperConfig` for this. I think that we can immediately have a compile error if the user used an unknown source property in `BeanMapping#ignoreUnmappedSourceProperties`.

I don't use this feature a lot, so I'm kind of surprised (or rather shocked 😄 ) that you can specify properties that don't exist in `BeanMapping#ignoreUnmappedSourceProperties`.. So I went ahead and tried by altering a testcase:

```java
package org.mapstruct.ap.test.source.ignore;

import org.mapstruct.BeanMapping;
import org.mapstruct.Mapper;
import org.mapstruct.ReportingPolicy;

@Mapper(
    unmappedTargetPolicy = ReportingPolicy.IGNORE,
    unmappedSourcePolicy = ReportingPolicy.ERROR
)
public interface IgnoreUnmappedSourceMapper {

    @BeanMapping(
        ignoreUnmappedSourceProperties = {
            "blah", // <<<<<<<<<<<<<<<<< not existing
            "name",
            "surname"
        }
    )
    PersonDto map(Person person);
}
```

And indeed.. you don't get  an error. 

To answer your question on my opinion @filiphr :
* So, no, we don't need another option. 
* And yes, please generate an error in this case.

Fully agreed upon!

**Comment by filiphr** [comment]
Thanks for looking into this as well @sjaakd. So @incaseoftrouble, please go ahead with what I said earlier, remove the option and automatically add a compile error if a defined source property in `BeanMapping#ignoreUnmappedSourceProperties` is unknown.

**Comment by incaseoftrouble** [comment]
Updated accordingly. Locally, integration tests (`./mvnw install`) succeed.

**Comment by filiphr** [comment]
Thanks for your work on this @incaseoftrouble
